dual/dual.go

```diff
@@ -0,0 +1,280 @@
+	wanCh := dht.WAN.FindProvidersAsync(reqCtx, key, count)
+	lanCh := dht.LAN.FindProvidersAsync(reqCtx, key, count)
+	go func() {
+		defer cancel()
+		defer close(outCh)
+
+		found := make(map[peer.ID]struct{}, count)
+		nch := 2
+		var pi peer.AddrInfo
+		for nch > 0 && count > 0 {
+			var ok bool
+			select {
+			case pi, ok = <-wanCh:
+				if !ok {
+					wanCh = nil
+					nch--
+					continue
+				}
+			case pi, ok = <-lanCh:
+				if !ok {
+					lanCh = nil
+					nch--
+					continue
+				}
+			}
+			// already found
+			if _, ok = found[pi.ID]; ok {
+				continue
+			}
+
+			select {
+			case outCh <- pi:
+				found[pi.ID] = struct{}{}
+				count--
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	return outCh
+}
+
+// FindPeer searches for a peer with given ID
+// Note: with signed peer records, we can change this to short circuit once either DHT returns.
+func (dht *DHT) FindPeer(ctx context.Context, pid peer.ID) (peer.AddrInfo, error) {
+	var wg sync.WaitGroup
+	wg.Add(2)
+	var wanInfo, lanInfo peer.AddrInfo
+	var wanErr, lanErr error
+	go func() {
+		defer wg.Done()
+		wanInfo, wanErr = dht.WAN.FindPeer(ctx, pid)
+	}()
+	go func() {
+		defer wg.Done()
+		lanInfo, lanErr = dht.LAN.FindPeer(ctx, pid)
+	}()
+
+	wg.Wait()
+
+	return peer.AddrInfo{
+		ID:    pid,
+		Addrs: append(wanInfo.Addrs, lanInfo.Addrs...),
+	}, mergeErrors(wanErr, lanErr)
+}
+
+func mergeErrors(a, b error) error {
+	if a == nil && b == nil {
+		return nil
+	} else if a != nil && b != nil {
+		return fmt.Errorf("%v, %v", a, b)
+	} else if a != nil {
+		return a
+	}
+	return b
+}
+
+// Bootstrap allows callers to hint to the routing system to get into a
+// Boostrapped state and remain there.
+func (dht *DHT) Bootstrap(ctx context.Context) error {
+	erra := dht.WAN.Bootstrap(ctx)
+	errb := dht.LAN.Bootstrap(ctx)
+	return mergeErrors(erra, errb)
+}
+
+// PutValue adds value corresponding to given Key.
+func (dht *DHT) PutValue(ctx context.Context, key string, val []byte, opts ...routing.Option) error {
+	if dht.activeWAN() {
+		return dht.WAN.PutValue(ctx, key, val, opts...)
+	}
+	return dht.LAN.PutValue(ctx, key, val, opts...)
+}
+
+// GetValue searches for the value corresponding to given Key.
+func (d *DHT) GetValue(ctx context.Context, key string, opts ...routing.Option) ([]byte, error) {
+	reqCtx, cncl := context.WithCancel(ctx)
+	defer cncl()
+
+	resChan := make(chan []byte)
+	defer close(resChan)
+	errChan := make(chan error)
+	defer close(errChan)
+	runner := func(impl *dht.IpfsDHT, valCh chan []byte, errCh chan error) {
+		val, err := impl.GetValue(reqCtx, key, opts...)
+		if err != nil {
+			errCh <- err
+			return
+		}
+		valCh <- val
+	}
+	go runner(d.WAN, resChan, errChan)
+	go runner(d.LAN, resChan, errChan)
+
+	var err error
+	var val []byte
+	select {
+	case val = <-resChan:
+		cncl()
+	case err = <-errChan:
+	}
+
+	// Drain or wait for the slower runner
+	select {
+	case secondVal := <-resChan:
+		if val == nil {
+			val = secondVal
```

Comment:
1. We need to wait for _at least_ the public DHT to finish.
2. We should take the better value (using the selector).